prov/gni/src/gnix_msg.c

```diff
@@ -558,6 +728,315 @@ static int __gnix_rndzv_req(void *arg)
+	struct gnix_tx_descriptor *txd = NULL;
+	size_t recv_len, get_len, ct_size, send_cnt, recv_cnt, txd_cnt;
+	uint64_t recv_ptr = 0UL;
+	/* TODO: Should this be the sender's rndzv thresh instead? */
+	size_t rndzv_thresh = ep->domain->params.msg_rendezvous_thresh;
+	gni_ct_get_post_descriptor_t *cur_ct = NULL;
+	void **next_ct = NULL;
+	uint32_t max_ct_size = nic->gni_cdm_modes & GNI_CDM_MODE_FMA_SHARED ?
+		1024*1024 : 1<<30;
+
+	GNIX_TRACE(FI_LOG_EP_DATA, "\n");
+
+	slist_init(&req->iov_txd_slist);
+	txd_cnt = ct_size = 0;
+	send_cnt = req->msg.send_iov_cnt;
+	recv_ptr = req->msg.recv_info[0].recv_addr;
+	recv_len = req->msg.recv_info[0].recv_len;
+	recv_cnt = req->msg.recv_iov_cnt;
+	use_tx_cq_blk = (ep->domain->data_progress == FI_PROGRESS_AUTO);
+
+	GNIX_DEBUG(FI_LOG_EP_DATA, "send_cnt = %lu, recv_cnt = %lu\n",
+		   send_cnt, recv_cnt);
+
+	/* Ensure the user's recv buffer is registered for recv/recvv */
+	if (!req->msg.recv_md[0]) {
+		struct fid_mr *auto_mr;
+
+		for (i = 0; i < recv_cnt; i++) {
+			auto_mr = NULL;
+			ret = gnix_mr_reg(&ep->domain->domain_fid.fid,
+					  (void *)
+					  req->msg.recv_info[i].recv_addr,
+					  req->msg.recv_info[i].recv_len,
+					  FI_READ | FI_WRITE, 0, 0, 0,
+					  &auto_mr, NULL);
+
+			if (ret != FI_SUCCESS) {
+				GNIX_DEBUG(FI_LOG_EP_DATA,
+					   "Failed to auto-register"
+					   " local buffer: %s\n",
+					   fi_strerror(-ret));
+
+				for (i--; i >= 0; i--) {
+					fi_close(&req->msg.recv_md[i]->mr_fid.fid);
+				}
+
+				return ret;
+			}
+
+			req->msg.recv_md[i] = container_of(
+				(void *) auto_mr,
+				struct gnix_fid_mem_desc,
+				mr_fid);
+
+			req->msg.recv_info[i].mem_hndl =
+				req->msg.recv_md[i]->mem_hndl;
+
+			GNIX_DEBUG(FI_LOG_EP_DATA, "auto-reg MR: %p\n",
+				   req->msg.recv_md[i]);
+
+		}
+		req->msg.recv_flags |= FI_LOCAL_MR;
+	}
+
+	/*
+	 * TODO: "coalesce" the user's sendv buffer before sending the control
+	 * message; do "the same" in _gnix_recvv.
+	 */
+	i = 0;
+
+	/* Iterate through the buffers and build the Fma and Rdma requests! */
+	while (i < send_cnt) {
+		get_len = MIN(recv_len, req->msg.send_info[i].send_len);
+
+		GNIX_DEBUG(FI_LOG_EP_DATA, "send_info[%d].send_len = %lu,"
+			   " recv_len = %lu, get_len = %lu\n", i,
+			   req->msg.send_info[i].send_len, recv_len, get_len);
+
+		if (get_len >= rndzv_thresh) { /* Build the rdma txd */
+			ret = _gnix_nic_tx_alloc(nic, &txd);
+
+			if (ret != FI_SUCCESS) {
+				/* We'll try again. */
+				GNIX_INFO(FI_LOG_EP_DATA, "_gnix_nic_tx_alloc()"
+					  " returned %s\n",
+					  fi_strerror(-ret));
+				return -FI_ENOSPC;
+			}
+
+			txd->completer_fn = __gnix_rndzv_iov_req_complete;
+			txd->req = req;
+			txd->gni_desc.type = GNI_POST_RDMA_GET;
+			txd->gni_desc.cq_mode = GNI_CQMODE_GLOBAL_EVENT;
+			txd->gni_desc.dlvr_mode = GNI_DLVMODE_PERFORMANCE;
+			txd->gni_desc.local_mem_hndl =
+				req->msg.recv_info[j].mem_hndl;
+			/* TODO: If the user passes in a incorrect mdh can we
+			 * report a useful error? */
```

Comment:
Not sure how you can tell, but if it's invalid you will get an error from ugni later.